add dkg round2 function

hughy opened the pull request:
defines round2 to take a round1 SecretPackage and a vector of round1 PublicPackages and run round2 of DKG.

finds the PublicPackage corresponding to self_identity and uses it to verify the checksums contained in all of the other PublicPackages.

iterates over round1 PublicPackages to construct map from identifier to frost package required for frost part2 inputs and to collect group_secret_key_shards

moves 'Error' enum from round1 to separate error module to allow reuse across dkg rounds, adds ChecksumError to enum

Commit:
add dkg round2 function

defines round2 to take a round1 SecretPackage and a vector of round1
PublicPackages and run round2 of DKG.

finds the PublicPackage corresponding to self_identity and uses it to verify the
checksums contained in all of the other PublicPackages.

iterates over round1 PublicPackages to construct map from identifier to frost
package required for frost part2 inputs and to collect group_secret_key_shards

moves 'Error' enum from round1 to separate error module to allow reuse across
dkg rounds, adds ChecksumError to enum

Commit:
adds missing license header in error.rs

andiflabs commented on src/dkg/round1.rs:
```diff
@@ -239,6 +240,20 @@ impl PublicPackage {
     }
 }
 
+impl Ord for PublicPackage {
+    #[inline]
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        Ord::cmp(&self.identity(), &other.identity())
+    }
+}
+
+impl PartialOrd<Self> for PublicPackage {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
```
This is comparing the identity only, which can lead to situations where one has 2 different packages generated from the same identity, but with different contents, which compare as equal.

Would it be better to instead use `sort_unstable_by` in the `round2` code?

andiflabs replied:
Of course I meant [`sort_unstable_by_key`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable_by_key), not `sort_unstable_by`

hughy replied:
That's a good point

Updated to use `sort_unstable_by_key`, since for now that's the only place that we need any ordering on PublicPackage

andiflabs commented on src/dkg/round2.rs:
```diff
@@ -118,9 +127,66 @@ pub fn import_secret_package(
 
+pub fn round2<'a, P, R: RngCore + CryptoRng>(
+    self_identity: &participant::Identity,
+    round1_secret_package: round1::SecretPackage,
```
Nit: can we use a `&round1::SecretPackage` instead, to avoid passing large values on the stack?

andiflabs commented on src/dkg/round2.rs:
```diff
@@ -118,9 +127,66 @@ pub fn import_secret_package(
+
+    for public_package in round1_public_packages {
+        if public_package.checksum() != self_public_package.checksum() {
+            return Err(Error::ChecksumError(ChecksumError));
```
Can we provide a more error message? I don't remember how we do it for TKD (if the error message is generated on ironfish-frost, ironfish-rust, ironfish-rust-nodejs, or typescript), but ideally we should use the same pattern.

This doesn't need to be done in this PR though.

hughy replied:
I'll update this in a separate PR since it will affect signing_commitment too

Commit:
uses sort_unstable_by for sorting round1 PublicPackages

removes Ord and PartialOrd implementations for PublicPackage that only compare
identities

Commit:
uses sort_unstable_by_key instead of sort_unstable_by

Commit:
passes reference to SecretPackage to round2
